cow-shell.1
.TH "cow\-shell" 1 "2006 Jan 25" "cowdancer" "cowdancer"
.SH "NAME"
cow\-shell \- Start a copy-on-write session and invoke a shell.
.SH SYNOPSIS
.BI "cow-shell [" "commandline" "]"
.SH DESCRIPTION
.B cow\-shell
Executes specified command with
.B cowdancer
or a shell if no command-line option is given.

Inside the cowdancer session, files under the current directory
are protected in a way so that a write operation to an i-node
which existed at the point of running cow-shell
is not modified.

Useful for managing hard-linked source-trees and 
scratch-filesystems.

.SH "EXAMPLES"
.TP
.B "cp -al cowdancer/ cowdancer.new &&  cd cowdancer.new && cow-shell  "
Create a hardlinked tree, and cd into the newly created tree.
After invoking cow-shell inside that tree, a write operation 
against a file that is hardlinked with the original tree 
will be copied to a new file before write.
.TP
.B "cp -al cowdancer/ cowdancer.new &&  cd cowdancer.new && cow-shell debuild"
Tries to run dpkg-buildpackage on the new tree,
to avoid damage to the original tree.

Configure debuild to ignore
.B ".ilist"
files, through 
.B "-i" 
or 
.B "-I"
option.

.PP
.SH "FILES"
.TP
.B ".ilist"
Generated dynamically in the current directory.
The file contains the list of i-node numbers that should be 
protected from write operation.

.TP
.B "/usr/lib/cowdancer/libcowdancer.so"
The shared library used internally that overrides some functions 
that potentially write to files.
This library implements the COW feature and is preloaded with 
LD_PRELOAD.

.TP
.B "/bin/cp"
Used for copying.
Requires GNU option
.B "\-a"
to be functional.

.SH "ENVIRONMENT"
.TP
.B "COWDANCER_ILISTFILE"
The path to the current 
.B .ilist
file.

.TP
.B "COWDANCER_IGNORE"
Used internally to work around infinite loops.
It will be unset.

.TP
.B "LD_PRELOAD"
Utilized to preload
.B libcowdancer.so

.TP
.B "COWDANCER_DEBUG"
Debugging option. 
Enables debug messages.
cowdancer will give verbose debug messages in standard error output.

.TP
.B "COWDANCER_SO"
Debugging option. 
Specifies a different path for libcowdancer.so.

.TP
.B "COWDANCER_REUSE=yes"
Enables reuse of cowdancer .ilist file found for consecutive runs of
cow-shell.

.SH "RETURN VALUE"

The functions that are overridden with cowdancer will give an errno
value of
.B ENOMEM
when there is a problem.

System isn't really out of memory, but this error code is chosen
because this error is usually handled gracefully by applications.

.SH "RESTRICTIONS"

Can only support directories that are scanned by 
.B cow-shell
command.

The operation is not semantically correct when files are originally
hardlinked, hardlinks are always broken.

Does not support anything when LD_PRELOAD trick does not work.

Does not work unless hardlinks are supported on the filesystem.

Not very user-friendly.

Leaves
.B ".ilist"
file around.

.SH "AUTHOR"
Junichi Uekawa (dancer@debian.org)

Upstream page is available at 
.B "http://www.netfort.gr.jp/~dancer/software/cowdancer.html"


.SH "SEE ALSO"
.BR "cowdancer-ilistcreate (" 1 ")"


NEWS
0.26 4 Mar 2007

+ add RPM spec file.

0.22 26 Aug 2006

+ cowbuilder documentation update.

0.21 20 Aug 2006

+ cowbuilder bugfix

0.15 29 May 2006

+ add help message to cowbuilder.
+ fix --basepath handling, --distribution handling
+ fix other option handling
+ update testsuite

0.14 24 May 2006

+ introducing an application "cowbuilder", a pbuilder frontend to use
  cowdancer.

0.13 5 Mar 2006

+ Fix: support quote and space in filenames
  (use exec instead of system.)

0.12 5 Mar 2006

+ Fix: support space in filenames. 
  Currently still doesn't support filename with "'", is there a need?

0.11 25 Jan 2006 

+ Check for error values on COW operation, and return error to application, with ENOMEM.

0.10 14 Jan 2006

+ fixes hangs with applications that are using pthread.
+ Added more checks in the testsuite.

0.9 8 Jan 2006

+ Change order of copying.  for the case when /lib/ld.so binary is
  written to; I cannot fork/exec shared binaries while ld.so does not
  exist.

0.8 28 Dec 2005

+ Speed improvement 1: save binary ilist in cow-shell and load the binary data 
  with mmap in cowdancer.
+ Speed improvement 2: sort ilist and do a binary search within it.
+ Make code a bit more thread-safe than it was.

0.7 23 Nov 2005

+ Bug fix: Work around 5.3.0 coreutils 'feature' which stat no longer 
  emits a newline for --format option.
+ Bug fix: Work around recent coreutils failing to do 
      cp -al 1/ 2/ 
  when 2/ does not already exist.
+ Environment variable COWDANCER_DEBUG can now be used for debugging
  e.g. COWDANCER_DEBUG=1 make -k check  (warning, log is massive)

0.6 28 Oct 2005

+ Add support for chown and chmod
+ Fix problem with powerpc, fill structs with memset.

0.5 21 Sep 2005

+ Support the case when SHELL is not a full path

0.4 30 Aug 2005

+ Fix Error checking so that it does more than kill(SIGSEGV) to self.

0.3 20 Aug 2005

+ Add regression testsuite.
+ Support symbollic links
+ fopen/fopen64 is now trapped
+ only COW regular files
+ do not try to cross device boundaries when creating the ilink list.
+ Now it is able to run pbuilder 0.130.

0.2 13 Aug 2005

+ Buffer overflow bug fixed.

0.1 13 Aug 2005

+ Initial Public Release.


cow-shell.c
/*BINFMTC: -DLIBDIR='"/usr/lib"' ilistcreate.c
  Copy-on-write filesystem invocation.

  GPL v2 or later
  Copyright 2005-2009 Junichi Uekawa.
 */
#define _GNU_SOURCE

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <string.h>
#include <sys/wait.h>
#include "ilist.h"

const char* ilist_PRGNAME="cow-shell";

/*
 * remove ilist after use.
 */
void ilist_deleter(const char* ilistfile)
{
  if(fork()==0)
    {
      /* I am the child process */

      pid_t parent_pid=getppid();
      if (daemon(0,1) < 0)
	{
	  perror("cow-shell daemon");
	  exit (-1);
	}
      while (kill(parent_pid,0) >= 0)
	{
	  sleep(1);
	}
      if (unlink(ilistfile)==-1)
	{
	  perror("cow-shell unlink .ilist");
	  exit(1);
	}
      exit(0);
    }
}


int main(int ac, char** av)
{
  /* give me a command-line to exec,
     and I will cow-keep what's under this directory. */
  const char* ilistpath="./.ilist";
  char*buf;
  struct stat st;
  int cowdancer_reuse;

  asprintf(&buf, "%s%s%s",
	   getenv("LD_PRELOAD")?:"",
	   getenv("LD_PRELOAD")?" ":"",
	   getenv("COWDANCER_SO") ?: LIBDIR "/cowdancer/libcowdancer.so");

  cowdancer_reuse=getenv("COWDANCER_REUSE") &&
    !strcmp(getenv("COWDANCER_REUSE"),"yes") ;

  if (cowdancer_reuse && !stat(ilistpath, &st))
    {
      /* if reuse flag is on and file already exists
	 do nothing */
    }
  else
    {
      if (unlink(ilistpath)==-1)
	{
	  if (errno == ENOENT)
	    {
	      /* expected */
	    }
	  else
	    {
	      perror("cow-shell: unlink of .ilist failed");
	      return 1;
	    }
	}
      if(ilistcreate(ilistpath, NULL))
	{
	  ilist_outofmemory(".ilist creation failed");
	  return 1;
	}
    }

  ilistpath=canonicalize_file_name(ilistpath);

  setenv("COWDANCER_ILISTFILE",
	  ilistpath,1);
  setenv("LD_PRELOAD",buf,1);
  unsetenv("COWDANCER_IGNORE");

  if (!cowdancer_reuse)
    {
      /* if reuse flag is not on, remove the file */
      ilist_deleter(ilistpath);
    }

  if (ac>1)
    execvp(av[1], av+1);
  else
    {
      const char* myshell=getenv("SHELL")?:"/bin/sh";
      fprintf(stderr, "Invoking %s\n", myshell);

      execlp(myshell,
	     myshell,
	     NULL);
      perror("cow-shell: exec");

      fprintf(stderr, "Falling back to /bin/sh\n");
      execlp("/bin/sh",
	     "/bin/sh",
	     NULL);
    }
  perror("cow-shell: exec");
  return 1;
}
